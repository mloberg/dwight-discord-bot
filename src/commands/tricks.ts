import { CommandBuilder, rollOption } from '../command';
import { FriendlyError } from '../error';
import { roll } from '../utils';

const bags = {
    gray: ['Weasel', 'Giant rat', 'Badger', 'Boar', 'Panther', 'Giant badger', 'Dire wolf', 'Giant elk'],
    rust: ['Rat', 'Owl', 'Mastiff', 'Goat', 'Giant goat', 'Giant boar', 'Lion', 'Brown bear'],
    tan: ['Jackal', 'Ape', 'Baboon', 'Axe beak', 'Black bear', 'Giant weasel', 'Giant hyena', 'Tiger'],
};

export default new CommandBuilder(async (command) => {
    const color = command.options.getString('color', true) as keyof typeof bags;
    const dice = command.options.getInteger('roll') || roll('d8');
    const result = bags[color][dice - 1];

    if (!result) {
        throw new FriendlyError('Unable to pull from the Bag of Tricks.');
    }

    await command.reply(result);
})
    .setName('tricks')
    .setDescription('Pull from a Bag of Tricks')
    .addStringOption((option) =>
        option
            .setName('color')
            .setDescription('The color of the Bag of Tricks')
            .setRequired(true)
            .addChoice('Gray', 'gray')
            .addChoice('Rust', 'rust')
            .addChoice('Tan', 'tan'),
    )
    .addIntegerOption(rollOption('d8'));
