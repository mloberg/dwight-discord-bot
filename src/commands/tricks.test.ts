import { CommandInteraction } from 'discord.js';

import { FriendlyError } from '../error';
import tricks from './tricks';

jest.mock('discord.js', () => ({
    CommandInteraction: jest.fn().mockImplementation(() => ({
        reply: jest.fn(),
        options: {
            getString: jest.fn(),
            getInteger: jest.fn(),
        },
    })),
}));

describe('/tricks', () => {
    it('is a slash command', () => {
        expect(tricks).toMatchSnapshot();
    });

    it('returns a creature', async () => {
        const command = jest.mocked(new CommandInteraction({} as never, {} as never), true);
        command.options.getString.mockReturnValue('rust');

        await tricks.handle(command);
        expect(command.reply).toHaveBeenCalledWith(expect.any(String));
    });

    it('returns a create for a dice roll', async () => {
        const command = jest.mocked(new CommandInteraction({} as never, {} as never), true);
        command.options.getString.mockReturnValue('gray');
        command.options.getInteger.mockReturnValue(3);

        await tricks.handle(command);
        expect(command.reply).toHaveBeenCalledWith('Badger');
    });

    it('throws an error if no creature is found', async () => {
        const command = jest.mocked(new CommandInteraction({} as never, {} as never), true);
        command.options.getString.mockReturnValue('tan');
        command.options.getInteger.mockReturnValue(20);

        await expect(tricks.handle(command)).rejects.toMatchError(
            new FriendlyError('Unable to pull from the Bag of Tricks.'),
        );
    });
});
