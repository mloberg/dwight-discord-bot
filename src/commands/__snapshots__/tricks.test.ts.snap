// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/tricks is a slash command 1`] = `
Object {
  "default_permission": undefined,
  "description": "Pull from a Bag of Tricks",
  "name": "tricks",
  "options": Array [
    Object {
      "choices": Array [
        Object {
          "name": "Gray",
          "value": "gray",
        },
        Object {
          "name": "Rust",
          "value": "rust",
        },
        Object {
          "name": "Tan",
          "value": "tan",
        },
      ],
      "description": "The color of the Bag of Tricks",
      "name": "color",
      "required": true,
      "type": 3,
    },
    Object {
      "choices": undefined,
      "description": "Value of a d8 roll.",
      "name": "roll",
      "required": false,
      "type": 4,
    },
  ],
}
`;
